readme.md
<span class="ModuleDoc YAMLDoc" id="QProgEdit" markdown="1">

# *module* QProgEdit

# QProgEdit

v3.0.0~pre2


QProgEdit is a PyQt4 widget that implements a full-featured text editor
component. It's primary target at the moment is
[OpenSesame](http://osdoc.cogsci.nl), a graphical experiment builder.

Copyright (2013-2015) Sebastiaan Mathôt

<http://www.cogsci.nl/smathot>

## Overview


- [Dependencies](#dependencies)
- [Example](#example)
- [class __QProgEdit.QEditor__](#class-__qprogeditqeditor__)
	- [function __QProgEdit\.QEditor\.\_\_init\_\___\(qProgEdit\)](#function-__qprogeditqeditor__init____qprogedit)
	- [function __QProgEdit\.QEditor\.applyCfg__\(\)](#function-__qprogeditqeditorapplycfg__)
	- [function __QProgEdit\.QEditor\.commentSelection__\(\)](#function-__qprogeditqeditorcommentselection__)
	- [function __QProgEdit\.QEditor\.cursorMoved__\(\)](#function-__qprogeditqeditorcursormoved__)
	- [function __QProgEdit\.QEditor\.focusInEvent__\(e\)](#function-__qprogeditqeditorfocusinevent__e)
	- [function __QProgEdit\.QEditor\.focusOutEvent__\(e\)](#function-__qprogeditqeditorfocusoutevent__e)
	- [function __QProgEdit\.QEditor\.highlightSelection__\(\)](#function-__qprogeditqeditorhighlightselection__)
	- [function __QProgEdit\.QEditor\.keyPressEvent__\(event\)](#function-__qprogeditqeditorkeypressevent__event)
	- [function __QProgEdit\.QEditor\.lang__\(\)](#function-__qprogeditqeditorlang__)
	- [function __QProgEdit\.QEditor\.onMarginClick__\(margin, line, state\)](#function-__qprogeditqeditoronmarginclick__margin-line-state)
	- [function __QProgEdit\.QEditor\.paste__\(\)](#function-__qprogeditqeditorpaste__)
	- [function __QProgEdit\.QEditor\.selectedText__\(currentLineFallback=False\)](#function-__qprogeditqeditorselectedtext__currentlinefallbackfalse)
	- [function __QProgEdit\.QEditor\.setKeyBindings__\(\)](#function-__qprogeditqeditorsetkeybindings__)
	- [function __QProgEdit\.QEditor\.setLang__\(lang=u'text'\)](#function-__qprogeditqeditorsetlang__langutext)
	- [function __QProgEdit\.QEditor\.setSymbolTree__\(symbolTree, symbolTreeWidgetItemClass=<class 'QProgEdit\.\_qsymboltreewidgetitem\.QSymbolTreeWidgetItem'>\)](#function-__qprogeditqeditorsetsymboltree__symboltree-symboltreewidgetitemclassclass-qprogedit_qsymboltreewidgetitemqsymboltreewidgetitem)
	- [function __QProgEdit\.QEditor\.setText__\(text\)](#function-__qprogeditqeditorsettext__text)
	- [function __QProgEdit\.QEditor\.symbols__\(\)](#function-__qprogeditqeditorsymbols__)
	- [function __QProgEdit\.QEditor\.text__\(\)](#function-__qprogeditqeditortext__)
	- [function __QProgEdit\.QEditor\.uncommentSelection__\(\)](#function-__qprogeditqeditoruncommentselection__)
	- [function __QProgEdit\.QEditor\.updateMarginWidth__\(\)](#function-__qprogeditqeditorupdatemarginwidth__)
	- [function __QProgEdit\.QEditor\.updateSymbolTree__\(\)](#function-__qprogeditqeditorupdatesymboltree__)
	- [function __QProgEdit\.QEditor\.validate__\(\)](#function-__qprogeditqeditorvalidate__)
	- [function __QProgEdit\.QEditor\.wheelEvent__\(event\)](#function-__qprogeditqeditorwheelevent__event)
- [class __QProgEdit.QEditorCfg__](#class-__qprogeditqeditorcfg__)
	- [function __QProgEdit\.QEditorCfg\.\_\_init\_\___\(parent=None\)](#function-__qprogeditqeditorcfg__init____parentnone)
	- [function __QProgEdit\.QEditorCfg\.version__\(\)](#function-__qprogeditqeditorcfgversion__)
- [class __QProgEdit.QEditorFind__](#class-__qprogeditqeditorfind__)
	- [function __QProgEdit\.QEditorFind\.\_\_init\_\___\(qProgEdit\)](#function-__qprogeditqeditorfind__init____qprogedit)
	- [function __QProgEdit\.QEditorFind\.caseSensitive__\(\)](#function-__qprogeditqeditorfindcasesensitive__)
	- [function __QProgEdit\.QEditorFind\.find__\(\)](#function-__qprogeditqeditorfindfind__)
	- [function __QProgEdit\.QEditorFind\.findText__\(\)](#function-__qprogeditqeditorfindfindtext__)
	- [function __QProgEdit\.QEditorFind\.loadUi__\(name\)](#function-__qprogeditqeditorfindloadui__name)
	- [function __QProgEdit\.QEditorFind\.lock__\(\)](#function-__qprogeditqeditorfindlock__)
	- [function __QProgEdit\.QEditorFind\.matchWhole__\(\)](#function-__qprogeditqeditorfindmatchwhole__)
	- [function __QProgEdit\.QEditorFind\.replace__\(\)](#function-__qprogeditqeditorfindreplace__)
	- [function __QProgEdit\.QEditorFind\.replaceAll__\(\)](#function-__qprogeditqeditorfindreplaceall__)
	- [function __QProgEdit\.QEditorFind\.replaceText__\(\)](#function-__qprogeditqeditorfindreplacetext__)
	- [function __QProgEdit\.QEditorFind\.setFindText__\(txt=u''\)](#function-__qprogeditqeditorfindsetfindtext__txtu)
	- [function __QProgEdit\.QEditorFind\.unlock__\(\)](#function-__qprogeditqeditorfindunlock__)
	- [function __QProgEdit\.QEditorFind\.unshow__\(\)](#function-__qprogeditqeditorfindunshow__)
- [class __QProgEdit.QEditorPrefs__](#class-__qprogeditqeditorprefs__)
	- [function __QProgEdit\.QEditorPrefs\.\_\_init\_\___\(qProgEdit\)](#function-__qprogeditqeditorprefs__init____qprogedit)
	- [function __QProgEdit\.QEditorPrefs\.apply__\(dummy=None\)](#function-__qprogeditqeditorprefsapply__dummynone)
	- [function __QProgEdit\.QEditorPrefs\.loadUi__\(name\)](#function-__qprogeditqeditorprefsloadui__name)
	- [function __QProgEdit\.QEditorPrefs\.refresh__\(\)](#function-__qprogeditqeditorprefsrefresh__)
- [class __QProgEdit.QEditorStatus__](#class-__qprogeditqeditorstatus__)
	- [function __QProgEdit\.QEditorStatus\.\_\_init\_\___\(qProgEdit\)](#function-__qprogeditqeditorstatus__init____qprogedit)
	- [function __QProgEdit\.QEditorStatus\.updateCursorPos__\(line=0, index=0\)](#function-__qprogeditqeditorstatusupdatecursorpos__line0-index0)
- [class __QProgEdit.QLangMenu__](#class-__qprogeditqlangmenu__)
	- [function __QProgEdit\.QLangMenu\.\_\_init\_\___\(tabCornerWidget\)](#function-__qprogeditqlangmenu__init____tabcornerwidget)
	- [function __QProgEdit\.QLangMenu\.setLang__\(action\)](#function-__qprogeditqlangmenusetlang__action)
- [function __QProgEdit\.QLexer__\(editor, lang=u'text', colorScheme=u'Default'\)](#function-__qprogeditqlexer__editor-langutext-colorschemeudefault)
- [class __QProgEdit.QProgEdit__](#class-__qprogeditqprogedit__)
	- [function __QProgEdit\.QProgEdit\.\_\_init\_\___\(tabManager, dPrint=None, title=u'Empty document', \*\*editorParams\)](#function-__qprogeditqprogedit__init____tabmanager-dprintnone-titleuempty-document-editorparams)
	- [function __QProgEdit\.QProgEdit\.dPrint__\(msg\)](#function-__qprogeditqprogeditdprint__msg)
	- [function __QProgEdit\.QProgEdit\.focusTab__\(\)](#function-__qprogeditqprogeditfocustab__)
	- [function __QProgEdit\.QProgEdit\.tabIndex__\(\)](#function-__qprogeditqprogedittabindex__)
	- [function __QProgEdit\.QProgEdit\.toggle__\(widget, visible\)](#function-__qprogeditqprogedittoggle__widget-visible)
	- [function __QProgEdit\.QProgEdit\.toggleFind__\(visible\)](#function-__qprogeditqprogedittogglefind__visible)
	- [function __QProgEdit\.QProgEdit\.togglePrefs__\(visible\)](#function-__qprogeditqprogedittoggleprefs__visible)
- [class __QProgEdit.QSymbolTreeWidgetItem__](#class-__qprogeditqsymboltreewidgetitem__)
	- [function __QProgEdit\.QSymbolTreeWidgetItem\.\_\_init\_\___\(editor, lineNo, \_type, name, argSpec\)](#function-__qprogeditqsymboltreewidgetitem__init____editor-lineno-_type-name-argspec)
	- [function __QProgEdit\.QSymbolTreeWidgetItem\.activate__\(\)](#function-__qprogeditqsymboltreewidgetitemactivate__)
- [class __QProgEdit.QTabCornerWidget__](#class-__qprogeditqtabcornerwidget__)
	- [function __QProgEdit\.QTabCornerWidget\.\_\_init\_\___\(tabManager, msg=None, handlerButtonText=None, runButton=False\)](#function-__qprogeditqtabcornerwidget__init____tabmanager-msgnone-handlerbuttontextnone-runbuttonfalse)
	- [function __QProgEdit\.QTabCornerWidget\.handlerButtonClicked__\(\)](#function-__qprogeditqtabcornerwidgethandlerbuttonclicked__)
	- [function __QProgEdit\.QTabCornerWidget\.update__\(\)](#function-__qprogeditqtabcornerwidgetupdate__)
- [class __QProgEdit.QTabManager__](#class-__qprogeditqtabmanager__)
	- [function __QProgEdit\.QTabManager\.\_\_init\_\___\(parent=None, cfg=<QProgEdit\.\_qeditorcfg\.QEditorCfg object at 0x7f73885d59d0>, tabsClosable=False, tabsMovable=False, msg=None, handlerButtonText=None, runButton=False\)](#function-__qprogeditqtabmanager__init____parentnone-cfgqprogedit_qeditorcfgqeditorcfg-object-at-0x7f73885d59d0-tabsclosablefalse-tabsmovablefalse-msgnone-handlerbuttontextnone-runbuttonfalse)
	- [function __QProgEdit\.QTabManager\.addTab__\(title=u'Empty document', select=True\)](#function-__qprogeditqtabmanageraddtab__titleuempty-document-selecttrue)
	- [function __QProgEdit\.QTabManager\.applyCfg__\(\)](#function-__qprogeditqtabmanagerapplycfg__)
	- [function __QProgEdit\.QTabManager\.closeTab__\(index=None\)](#function-__qprogeditqtabmanagerclosetab__indexnone)
	- [function __QProgEdit\.QTabManager\.isAnyModified__\(\)](#function-__qprogeditqtabmanagerisanymodified__)
	- [function __QProgEdit\.QTabManager\.runSelectedText__\(\)](#function-__qprogeditqtabmanagerrunselectedtext__)
	- [function __QProgEdit\.QTabManager\.runText__\(\)](#function-__qprogeditqtabmanagerruntext__)
	- [function __QProgEdit\.QTabManager\.selectTab__\(index\)](#function-__qprogeditqtabmanagerselecttab__index)
	- [function __QProgEdit\.QTabManager\.selectedText__\(index=None, currentLineFallback=False\)](#function-__qprogeditqtabmanagerselectedtext__indexnone-currentlinefallbackfalse)
	- [function __QProgEdit\.QTabManager\.setFocus__\(index=None\)](#function-__qprogeditqtabmanagersetfocus__indexnone)
	- [function __QProgEdit\.QTabManager\.setText__\(text, index=None\)](#function-__qprogeditqtabmanagersettext__text-indexnone)
	- [function __QProgEdit\.QTabManager\.switchTabLeft__\(\)](#function-__qprogeditqtabmanagerswitchtableft__)
	- [function __QProgEdit\.QTabManager\.switchTabRight__\(\)](#function-__qprogeditqtabmanagerswitchtabright__)
	- [function __QProgEdit\.QTabManager\.tab__\(index=None\)](#function-__qprogeditqtabmanagertab__indexnone)
	- [function __QProgEdit\.QTabManager\.tabChanged__\(index\)](#function-__qprogeditqtabmanagertabchanged__index)
	- [function __QProgEdit\.QTabManager\.tabIndex__\(index=None\)](#function-__qprogeditqtabmanagertabindex__indexnone)
	- [function __QProgEdit\.QTabManager\.tabs__\(\)](#function-__qprogeditqtabmanagertabs__)
	- [function __QProgEdit\.QTabManager\.text__\(index=None\)](#function-__qprogeditqtabmanagertext__indexnone)
	- [function __QProgEdit\.QTabManager\.toggleFind__\(visible\)](#function-__qprogeditqtabmanagertogglefind__visible)
	- [function __QProgEdit\.QTabManager\.togglePrefs__\(visible\)](#function-__qprogeditqtabmanagertoggleprefs__visible)
- [class __QProgEdit.QUiLoader__](#class-__qprogeditquiloader__)
	- [function __QProgEdit\.QUiLoader\.loadUi__\(name\)](#function-__qprogeditquiloaderloadui__name)



## Dependencies

- `PyQt4`
- `Qscintilla2`

## Example

~~~ python
#!/usr/bin/env python
#-*- coding:utf-8 -*-

"""
This file is part of QProgEdit.

QProgEdit is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

QProgEdit is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with QProgEdit.  If not, see <http://www.gnu.org/licenses/>.
"""

import os
import sys
from QProgEdit.qt import QtGui, QtCore
from QProgEdit import QTabManager, validate

def cursorRowChanged(index, rowFrom, rowTo):

	print(u'curorRowChanged(): %d, %d, %d' % (index, rowFrom, rowTo))

def focusLost(index):

	print(u'focusOut(): %s' % index)

def focusReceived(index):

	print(u'focusReceived(): %s' % index)

def handlerButtonClicked(index):

	print(u'handlerButtonClicked(): %s' % index)

def activateSymbolTree(treeWidgetItem):

	if hasattr(treeWidgetItem, u'activate'):
		treeWidgetItem.activate()

def runSelectedText(s):

	print('run:\n%s' % s)

def main():

	"""Runs a simple QProgEdit demonstration."""

	validate.addPythonBuiltins(['builtin_var'])
	app = QtGui.QApplication(sys.argv)

	treeWidgetItem1 = QtGui.QTreeWidgetItem([u'Tab 1'])
	treeWidgetItem3 = QtGui.QTreeWidgetItem([u'Tab 3'])
	symbolTree = QtGui.QTreeWidget()
	symbolTree.addTopLevelItem(treeWidgetItem1)
	symbolTree.addTopLevelItem(treeWidgetItem3)
	symbolTree.itemActivated.connect(activateSymbolTree)

	tabManager = QTabManager(handlerButtonText=u'apply', runButton=True)
	tabManager.setWindowIcon(QtGui.QIcon.fromTheme(u'accessories-text-editor'))
	tabManager.setWindowTitle(u'QProgEdit')
	tabManager.resize(800, 600)

	tabManager.cursorRowChanged.connect(cursorRowChanged)
	tabManager.focusLost.connect(focusLost)
	tabManager.focusReceived.connect(focusReceived)
	tabManager.handlerButtonClicked.connect(handlerButtonClicked)
	tabManager.execute.connect(runSelectedText)

	tab = tabManager.addTab(u'Tab 1')
	tab.setLang(u'Python')
	tab.setSymbolTree(treeWidgetItem1)
	tab.setText(open(__file__).read())

	tab = tabManager.addTab(u'Tab 2')
	tab.setText(u'Some plain text')

	tab = tabManager.addTab(u'Tab 3')
	tab.setLang(u'Python')
	tab.setSymbolTree(treeWidgetItem3)
	if os.path.exists(u'content.txt'):
		tab.setText(open(u'content.txt').read())

	layout = QtGui.QHBoxLayout()
	layout.addWidget(symbolTree)
	layout.addWidget(tabManager)
	container = QtGui.QWidget()
	container.setLayout(layout)
	container.show()

	res = app.exec_()
	open(u'content.txt', u'w').write(tab.text())
	sys.exit(res)

if __name__ == '__main__':
	main()

~~~

<span class="ClassDoc YAMLDoc" id="QProgEdit-QEditor" markdown="1">

## class __QProgEdit.QEditor__

A single editor widget, which is embedded in a QProgEdit widget.

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-__init__" markdown="1">

### function __QProgEdit\.QEditor\.\_\_init\_\___\(qProgEdit\)

Constructor.

__Arguments:__

- `qProgEdit` -- The parent QProgEdit.
	- Type: QProgEdit

</span>

[QProgEdit.QEditor.__init__]: #QProgEdit-QEditor-__init__
[QEditor.__init__]: #QProgEdit-QEditor-__init__
[__init__]: #QProgEdit-QEditor-__init__

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-applyCfg" markdown="1">

### function __QProgEdit\.QEditor\.applyCfg__\(\)

Applies the configuration.

</span>

[QProgEdit.QEditor.applyCfg]: #QProgEdit-QEditor-applyCfg
[QEditor.applyCfg]: #QProgEdit-QEditor-applyCfg
[applyCfg]: #QProgEdit-QEditor-applyCfg

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-commentSelection" markdown="1">

### function __QProgEdit\.QEditor\.commentSelection__\(\)

Comments out the currently selected text.

</span>

[QProgEdit.QEditor.commentSelection]: #QProgEdit-QEditor-commentSelection
[QEditor.commentSelection]: #QProgEdit-QEditor-commentSelection
[commentSelection]: #QProgEdit-QEditor-commentSelection

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-cursorMoved" markdown="1">

### function __QProgEdit\.QEditor\.cursorMoved__\(\)

Is called whenever the cursor moves, checks whether the cursor has jumped from one line to the next, and, if so, calls the relevant functions.

</span>

[QProgEdit.QEditor.cursorMoved]: #QProgEdit-QEditor-cursorMoved
[QEditor.cursorMoved]: #QProgEdit-QEditor-cursorMoved
[cursorMoved]: #QProgEdit-QEditor-cursorMoved

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-focusInEvent" markdown="1">

### function __QProgEdit\.QEditor\.focusInEvent__\(e\)

Called when the editor receives focus.

__Arguments:__

- `e` -- No description

</span>

[QProgEdit.QEditor.focusInEvent]: #QProgEdit-QEditor-focusInEvent
[QEditor.focusInEvent]: #QProgEdit-QEditor-focusInEvent
[focusInEvent]: #QProgEdit-QEditor-focusInEvent

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-focusOutEvent" markdown="1">

### function __QProgEdit\.QEditor\.focusOutEvent__\(e\)

Called when the editor loses focus.

__Arguments:__

- `e` -- No description

</span>

[QProgEdit.QEditor.focusOutEvent]: #QProgEdit-QEditor-focusOutEvent
[QEditor.focusOutEvent]: #QProgEdit-QEditor-focusOutEvent
[focusOutEvent]: #QProgEdit-QEditor-focusOutEvent

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-highlightSelection" markdown="1">

### function __QProgEdit\.QEditor\.highlightSelection__\(\)

Highlights all parts of the text that match the current selection.

</span>

[QProgEdit.QEditor.highlightSelection]: #QProgEdit-QEditor-highlightSelection
[QEditor.highlightSelection]: #QProgEdit-QEditor-highlightSelection
[highlightSelection]: #QProgEdit-QEditor-highlightSelection

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-keyPressEvent" markdown="1">

### function __QProgEdit\.QEditor\.keyPressEvent__\(event\)

Intercepts certain keypress events to implement custom copy-pasting and zooming.

__Arguments:__

- `event` -- No description
	- Type: QKeyPressEvent

</span>

[QProgEdit.QEditor.keyPressEvent]: #QProgEdit-QEditor-keyPressEvent
[QEditor.keyPressEvent]: #QProgEdit-QEditor-keyPressEvent
[keyPressEvent]: #QProgEdit-QEditor-keyPressEvent

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-lang" markdown="1">

### function __QProgEdit\.QEditor\.lang__\(\)

No description specified.

__Returns:__

The language of the editor.

- Type: unicode

</span>

[QProgEdit.QEditor.lang]: #QProgEdit-QEditor-lang
[QEditor.lang]: #QProgEdit-QEditor-lang
[lang]: #QProgEdit-QEditor-lang

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-onMarginClick" markdown="1">

### function __QProgEdit\.QEditor\.onMarginClick__\(margin, line, state\)

Shows validation errors when the margin symbol is clicked.

__Arguments:__

- `margin` -- The margin number.
	- Type: int
- `line` -- The line number.
	- Type: int
- `state` -- The keyboard state.
	- Type: int

</span>

[QProgEdit.QEditor.onMarginClick]: #QProgEdit-QEditor-onMarginClick
[QEditor.onMarginClick]: #QProgEdit-QEditor-onMarginClick
[onMarginClick]: #QProgEdit-QEditor-onMarginClick

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-paste" markdown="1">

### function __QProgEdit\.QEditor\.paste__\(\)

Re-implements the paste method to allow modification of paste content.

</span>

[QProgEdit.QEditor.paste]: #QProgEdit-QEditor-paste
[QEditor.paste]: #QProgEdit-QEditor-paste
[paste]: #QProgEdit-QEditor-paste

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-selectedText" markdown="1">

### function __QProgEdit\.QEditor\.selectedText__\(currentLineFallback=False\)

Returns the selected text.

__Keywords:__

- `currentLineFallback` -- Indicates whether the current line should be returned if no text has been selected. Otherwise, an empty string is returned.
	- Type: bool
	- Default: False

__Returns:__

The selected text.

- Type: str

</span>

[QProgEdit.QEditor.selectedText]: #QProgEdit-QEditor-selectedText
[QEditor.selectedText]: #QProgEdit-QEditor-selectedText
[selectedText]: #QProgEdit-QEditor-selectedText

<span class="FunctionDoc YAMLDoc" id="QProgEdit-QEditor-setKeyBindings" markdown="1">


debian/changelog
python-qprogedit (3.2.2-ubuntu1) wily; urgency=medium

  * Add more Python built-ins

 -- Sebastiaan Mathot <sebastiaan@mantis>  Thu, 07 Jan 2016 12:01:30 +0100

python-qprogedit (3.2.0-ubuntu2) wily; urgency=medium

  * Add setInvalid() function to manually invalidate lines

 -- Sebastiaan Mathot <sebastiaan@mantis>  Wed, 28 Oct 2015 16:34:08 +0100

python-qprogedit (3.1.0-ubuntu1) vivid; urgency=medium

  * Add a PyFlakes filter to ignore unwanted messages

 -- Sebastiaan Mathot <sebastiaan@mantis>  Wed, 21 Oct 2015 11:05:02 +0200

python-qprogedit (3.0.1release-ubuntu1) vivid; urgency=medium

  * Python 3 support
  * PyQt5 compatibility
  * Monokai color theme 

 -- Sebastiaan Mathot <sebastiaan@mantis>  Mon, 05 Oct 2015 11:04:31 +0200

python-qprogedit (2.1.0-ubuntu2) trusty; urgency=medium

  * Add OpenSesame-script lexer
  * Improve keywords in Python lexer

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Sun, 09 Nov 2014 13:23:03 +0100

python-qprogedit (2.0.5-ubuntu1) trusty; urgency=medium

  * Improve focus behavior
  * Add new colorschemes and add support for bold and italics

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 04 Nov 2014 12:24:13 +0100

python-qprogedit (2.0.4-ubuntu1) trusty; urgency=medium

  * Always rebuild symbol tree when a new symbol tree is set

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Thu, 25 Sep 2014 18:40:11 +0200

python-qprogedit (2.0.3-ubuntu1) trusty; urgency=medium

  * Fix crash when changing language through language menu

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 23 Sep 2014 15:15:40 +0200

python-qprogedit (2.0.2-ubuntu1) trusty; urgency=medium

  * Avoid (more) shortcut clashes when multiple instances are visible

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 09 Sep 2014 17:22:25 +0200

python-qprogedit (2.0.1-ubuntu1) trusty; urgency=medium

  * Catch multiline defs in Python symbols
  * Avoid shortcut clashes when multiple instances are visible

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Mon, 08 Sep 2014 21:08:38 +0200

python-qprogedit (2.0.0-ubuntu1) trusty; urgency=medium

  * Not backwards compatible with 1.X
  * Simplified and more conistent API
  * Compatible with Python 3
  * Add symbol-tree browser
  * Use yamldoc-style docstrings
  * Use signals instead of callback functions
  * Dynamically load UI

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Sat, 16 Aug 2014 18:27:20 +0200

python-qprogedit (1.4.0-ubuntu1) trusty; urgency=medium

  * Add symbol-tree browser
  * Add cursorRowChanged signal

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Fri, 15 Aug 2014 14:27:14 +0200

python-qprogedit (1.3.3-ubuntu1) trusty; urgency=medium

  * Fix a bug that caused interference between QtWebKit and QProgEdit

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Thu, 10 Jul 2014 10:12:07 +0200

python-qprogedit (1.3.2-ubuntu1) trusty; urgency=medium

  * Also update selection highlight when moving or typing

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Wed, 04 Jun 2014 18:37:37 +0200

python-qprogedit (1.3.1-ubuntu1) trusty; urgency=medium

  * Implement tab switching by pressing Alt+Left/ Alt+Right
  * Fix interference between selection highlighting and find/ replace

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Wed, 14 May 2014 14:14:36 +0200

python-qprogedit (1.3.0-ubuntu2) trusty; urgency=medium

  * Improved find/ replace behavior
  * Highlight text that matches selection

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 13 May 2014 16:27:20 +0200

python-qprogedit (1.2.4-ubuntu1) trusty; urgency=medium

  * Place cursor after inserted text when pasting (#11)
  * Improve focus behavior and tab switching

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 29 Apr 2014 11:57:01 +0200

python-qprogedit (1.2.3-ubuntu1) trusty; urgency=medium

  * Set EOL mode to linux to fix problems on OSX

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Fri, 25 Apr 2014 09:55:23 +0200

python-qprogedit (1.2.2-2) unstable; urgency=low

  * Do not accept popup-focus events

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Mon, 31 Mar 2014 14:02:33 +0200

python-qprogedit (1.2.1-ubuntu1) saucy; urgency=low

  * Don't assume that pyflakes accepts builtins keyword (#6)
  * Only call focusOut handler when isModified is True

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 11 Mar 2014 17:08:19 +0100

python-qprogedit (1.2.0-ubuntu1) saucy; urgency=low

  * Zoom on scroll

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Mon, 10 Mar 2014 16:38:20 +0100

python-qprogedit (1.1.0-ubuntu1) saucy; urgency=low

  * Auto-fix space-based indentation
  * Use Python lexer for OpenSesame script

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Mon, 03 Mar 2014 18:59:42 +0100

python-qprogedit (1.0.0-ubuntu1) saucy; urgency=low

  * Better implementation of focus-out handler

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Thu, 02 Jan 2014 17:23:26 +0100

python-qprogedit (0.3.1-ubuntu1) saucy; urgency=low

  * Add missing options to specify Python builtins to pyflakes

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Thu, 12 Dec 2013 17:11:22 +0100

python-qprogedit (0.3.0-ubuntu1) saucy; urgency=low

  * Add content validation using pyflakes
  * Do not get confused by capitalization in source language

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Thu, 12 Dec 2013 16:48:49 +0100

python-qprogedit (0.2.0-ubuntu1) saucy; urgency=low

  * Improved and configurable block (un)commenting

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Tue, 10 Dec 2013 12:36:31 +0100

python-qprogedit (0.1.4-ubuntu1) saucy; urgency=low

  * Do not crash when an empty string is specified as colorScheme

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Wed, 13 Nov 2013 18:10:43 +0100

python-qprogedit (0.1.3-ubuntu1) saucy; urgency=low

  * Focus find box when toggled

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Mon, 11 Nov 2013 18:02:01 +0100

python-qprogedit (0.1.2-ubuntu1) saucy; urgency=low

  * Minor bugfixes

 -- Sebastiaan Mathot <s.mathot@cogsci.nl>  Wed, 30 Oct 2013 14:54:18 +0100

python-qprogedit (0.1.1-ubuntu1) saucy; urgency=low


example.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

"""
This file is part of QProgEdit.

QProgEdit is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

QProgEdit is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with QProgEdit.  If not, see <http://www.gnu.org/licenses/>.
"""

import os
import sys
from QProgEdit.qt import QtGui, QtCore
from QProgEdit import QTabManager, validate

def cursorRowChanged(index, rowFrom, rowTo):

	print(u'curorRowChanged(): %d, %d, %d' % (index, rowFrom, rowTo))

def focusLost(index):

	print(u'focusOut(): %s' % index)

def focusReceived(index):

	print(u'focusReceived(): %s' % index)

def handlerButtonClicked(index):

	print(u'handlerButtonClicked(): %s' % index)

def activateSymbolTree(treeWidgetItem):

	if hasattr(treeWidgetItem, u'activate'):
		treeWidgetItem.activate()

def runSelectedText(s):

	print('run:\n%s' % s)

def main():

	"""Runs a simple QProgEdit demonstration."""

	validate.addPythonBuiltins(['builtin_var'])
	validate.setPyFlakesFilter(lambda msg: msg.message == u'undefined name %r')
	app = QtGui.QApplication(sys.argv)

	treeWidgetItem1 = QtGui.QTreeWidgetItem([u'Tab 1'])
	treeWidgetItem3 = QtGui.QTreeWidgetItem([u'Tab 3'])
	symbolTree = QtGui.QTreeWidget()
	symbolTree.addTopLevelItem(treeWidgetItem1)
	symbolTree.addTopLevelItem(treeWidgetItem3)
	symbolTree.itemActivated.connect(activateSymbolTree)

	tabManager = QTabManager(handlerButtonText=u'apply', runButton=True)
	tabManager.setWindowIcon(QtGui.QIcon.fromTheme(u'accessories-text-editor'))
	tabManager.setWindowTitle(u'QProgEdit')
	tabManager.resize(800, 600)

	tabManager.cursorRowChanged.connect(cursorRowChanged)
	tabManager.focusLost.connect(focusLost)
	tabManager.focusReceived.connect(focusReceived)
	tabManager.handlerButtonClicked.connect(handlerButtonClicked)
	tabManager.execute.connect(runSelectedText)

	tab = tabManager.addTab(u'Tab 1')
	tab.setLang(u'Python')
	tab.setSymbolTree(treeWidgetItem1)
	tab.setText(open(__file__).read())

	tab = tabManager.addTab(u'Tab 2')
	tab.setText(u'Some plain text')

	tab = tabManager.addTab(u'Tab 3')
	tab.setLang(u'Python')
	tab.setSymbolTree(treeWidgetItem3)
	if os.path.exists(u'content.txt'):
		tab.setText(open(u'content.txt').read())

	layout = QtGui.QHBoxLayout()
	layout.addWidget(symbolTree)
	layout.addWidget(tabManager)
	container = QtGui.QWidget()
	container.setLayout(layout)
	container.show()

	res = app.exec_()
	open(u'content.txt', u'w').write(tab.text())
	sys.exit(res)

if __name__ == '__main__':
	main()
